hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/container/UpgradeSubcommand.java

```diff
@@ -0,0 +1,194 @@
+
+  @CommandLine.Option(names = {"--volume"},
+      required = false,
+      description = "volume name")
+  private String volume;
+
+  @CommandLine.Option(names = {"-y", "--yes"},
+      description = "Continue without interactive user confirmation")
+  private boolean yes;
+
+  private static OzoneConfiguration ozoneConfiguration;
+
+
+  @Override
+  public Void call() throws Exception {
+    final UpgradeChecker upgradeChecker = new UpgradeChecker();
+
+    OzoneConfiguration configuration = getConfiguration();
+
+    Pair<Boolean, String> pair =
+        upgradeChecker.checkDatanodeNotStarted();
+    final boolean dnNotStarted = pair.getKey();
+    if (!dnNotStarted) {
+      out().println(pair.getValue());
+      return null;
+    }
+
+    final Pair<Integer, Integer> layoutVersion =
+        upgradeChecker.getLayoutVersion(configuration);
+    final Integer softwareLayoutVersion = layoutVersion.getLeft();
+    final Integer metadataLayoutVersion = layoutVersion.getRight();
+    final int needLayoutVersion =
+        HDDSLayoutFeature.DATANODE_SCHEMA_V3.layoutVersion();
+
+    if (metadataLayoutVersion < needLayoutVersion ||
+        softwareLayoutVersion < needLayoutVersion) {
+      out().println(String.format(
+          "Please upgrade your software version, no less than %s," +
+              " current metadata layout version is %d," +
+              " software layout version is %d",
+          HDDSLayoutFeature.DATANODE_SCHEMA_V3.name(),
+          metadataLayoutVersion, softwareLayoutVersion));
+      return null;
+    }
+
+    if (!Strings.isNullOrEmpty(volume)) {
+      configuration.set(ScmConfigKeys.HDDS_DATANODE_DIR_KEY, volume);
+    }
+
+    DatanodeDetails dnDetail =
+        UpgradeUtils.getDatanodeDetails(configuration);
+
+    final HddsProtos.NodeOperationalState opState =
+        dnDetail.getPersistedOpState();
+
+    if (!opState.equals(HddsProtos.NodeOperationalState.IN_MAINTENANCE)) {
+      out().println("This command requires the datanode's " +
+          "NodeOperationalState to be IN_MAINTENANCE, currently is " +
+          opState);
+      return null;
+    }
+
+    List<HddsVolume> allVolume =
+        upgradeChecker.getAllVolume(dnDetail, configuration);
+
+    final List<File> volumeDBPath = upgradeChecker.getVolumeDBPath(allVolume);
+
+    if (volumeDBPath.isEmpty()) {
+      out().println("No volume db store exists, need finalizae data node.");
+      return null;
+    }
+
+    for (HddsVolume hddsVolume : allVolume) {
+      if (UpgradeChecker.checkAlreadyMigrate(hddsVolume)) {
+        out().println("Volume " + hddsVolume.getVolumeRootDir() +
+            " it's already upgraded, skip it.");
+        return null;
+      }
+    }
+
+    if (!yes) {
+      Scanner scanner = new Scanner(new InputStreamReader(
+          System.in, StandardCharsets.UTF_8));
+      System.out.println(
+          "The volume db store, will be automatically backup," +
+              " should you continue to upgrade?");
```

Comment:
"The volume db store, will be automatically backup," +
              " should you continue to upgrade?"   ->

"The volume db store(s), will be automatically backup. " + 
              "Enter 'yes' to continue: "